### DISCUSSION TITLE
[Bug or Feature?] Top-Level Await Dynamic Import Incompatibility with Provide Plugin Causing Undefined Variables

### DISCUSSION THREAD

**Discussion by Kane-Kuroneko** [webpack/webpack#19007] [Q&A]
# Bug report

**What is the current behavior?**
All variables exported from an ESM that contains a top-level awaited dynamic import are undefined.
**If the current behavior is a bug, please provide the steps to reproduce.**
```js
if(isElectron){
   var { IPC } = await import('../../ENV/electron')
}
export const something = () => {
   IPC?.on('json',() => {...})
}
```
This case demonstrates that some logic depends on a dynamic module, and `something` is provided by ProvidePlugin, but it's `undefined` at runtime. This issue can be resolved by replacing `await import()` with `require()`, but this makes the ESM code impure.

**What is the expected behavior?**
Expect the provide-plugin to work correctly when a variable is exported from an ESM that contains a top-level awaited dynamic import.

**Other relevant information:**
webpack version:  ^5.95.0
Node.js version: 23.1.0
Operating System: win11
Additional tools:

**Comment by Callme-VR** [comment]
just try to use await sync
require()

**Comment by Kane-Kuroneko** [comment]
> just try to use await sync require()

I know , but I suppose "This issue can be resolved by replacing await import() with require(), but this makes the ESM code impure."

**Comment by alexander-akait** [comment]
Unfortunately I can't help without a real example, it should work and we have tests on such cases, I'll temporarily move it to discussions, as soon as you provide an example we will try to understand why it doesn't work in your case

**Reply by Kane-Kuroneko** [reply]
I created a repo for this case. You can see the issue once you touch src/thing.ts. 
Clicking count on the page will print the value of thing, which was provided.

here is the repo : https://github.com/Kane-Kuroneko/test_case-webpack-ProvidePlugin-TopLevel-Await

**Reply by Kane-Kuroneko** [reply]
> Unfortunately I can't help without a real example, it should work and we have tests on such cases, I'll temporarily move it to discussions, as soon as you provide an example we will try to understand why it doesn't work in your case

Any response plz?

**Reply by alexander-akait** [reply]
I see your problem, unfortunately it is a limitation, the main problem is why it is difficult to implement - you use the variable synchronously, although the result should be `awaited`, it means the parent module should also become asynchronous, i.e. like we have top level await in parent module, this may break execution semantics in some cases, technically it is actually implementable, but I would recommend not to do it - using such values ​​you will get a loss of performance in runtime, I would recommend calculating such values ​​only when it is really required.

Sorry for the delay in replying, feel free to feedback.
